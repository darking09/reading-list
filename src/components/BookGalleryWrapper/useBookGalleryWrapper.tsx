import { useState, useEffect, useTransition } from "react";
import { Book } from '@typesFiles/Books'
import { updateToReadingList } from "@api/index"

export default function useBookGalleryWrapper(book: Book) {
  let [isPending, startTransition] = useTransition()
  useEffect(() => {
    setIsAvailable(checkIsAvailability(book));
  }, [book]);

  //TODO: Add loading function to show loading state
  /*useEffect(() => {
    console.log(isPending)
  }, [isPending]);*/

  const checkIsAvailability = (b: Book) => {
    return b.book.available ?? true
  }

  const [isAvailable, setIsAvailable] = useState(checkIsAvailability(book));

  const onClick = () => {
    let newBook = Object.assign({}, book)

    if (!newBook.book?.available) {
      newBook = Object.assign(book, { book: { ...book.book, available: true } })
    }

    newBook.book.available = false

    startTransition(() => updateToReadingList(newBook))
  }

  return {
    onClick,
    isAvailable
  };
}

