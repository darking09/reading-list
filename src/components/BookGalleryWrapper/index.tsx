'use client'
import { BookGalleryWrapperProps } from '@typesFiles/props/bookGalleryWrapper'
import useBookGalleryWrapper from './useBookGalleryWrapper'

export default function BookGalleryWrapper({ children, book } : BookGalleryWrapperProps) {
  const { isAvailable, onClick } = useBookGalleryWrapper(book)

  if (isAvailable) {
    return (
      <div className="cursor-pointer" onClick={onClick}>
        {children}
      </div>
    )
  }

  return (
    <div className="blur-sm contrast-10 opacity-80">
      {children}
    </div>
  )
}
