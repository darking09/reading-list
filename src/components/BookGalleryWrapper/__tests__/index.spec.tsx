import BookGalleryWrapper from "..";
import { render, screen } from "@testing-library/react";
import '@testing-library/jest-dom';

jest.mock('@components/BookGalleryWrapper/useBookGalleryWrapper', () => () => ({
  isAvailable: true,
  onClick: jest.fn(),
}))

describe('BookGalleryWrapper', () => {
  it('should render', () => {
    const { container } = render(
      <BookGalleryWrapper
        book={{
          id: 1,
          pages: 1,
          genre_id: 1,
          author_id: {
            name: 'name',
            otherBooks: [],
          },
          synopsis: 'synopsis',
          year: 1,
          ISBN: 'ISBN',
          is_available: true,
          title: 'title',
          cover: 'cover',
          created_at: new Date(),
        }}
      >
        <div>children</div>
      </BookGalleryWrapper>
    )
    expect(container).toMatchSnapshot()
  })

  it('should render a book with a cover', () => {
    render(
      <BookGalleryWrapper
        book={{
          id: 1,
          pages: 1,
          genre_id: 1,
          author_id: {
            name: 'name',
            otherBooks: [],
          },
          synopsis: 'synopsis',
          year: 1,
          ISBN: 'ISBN',
          is_available: true,
          title: 'title',
          cover: 'cover',
          created_at: new Date(),
        }}
      >
        <div>children</div>
      </BookGalleryWrapper>
    )

    const element = screen.getByText(/children/)
    expect(element).toBeInTheDocument()
  })
})
