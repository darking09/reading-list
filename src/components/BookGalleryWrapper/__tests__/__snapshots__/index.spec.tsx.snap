// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookGalleryWrapper should render 1`] = `
<div>
  <div
    class="cursor-pointer"
  >
    <div>
      children
    </div>
  </div>
</div>
`;
